libs/ledger/include/ledger/chain/time_travelogue.hpp

```diff
@@ -0,0 +1,54 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "chain/constants.hpp"
+#include "core/serializers/base_types.hpp"
+#include "ledger/chain/block.hpp"
+
+namespace fetch {
+namespace ledger {
+
+/**
+ * Packet format used to convey heaviest chain for node sync.
+ */
+template <class B>
```

Comment:
Can we remove the template on this/make it concrete?

Reply:
> Can we remove the template on this/make it concrete?

Then one minor issue arises. On the receiving end, when you cast `MainChainRpcService` to `Travelogue`, the blocks it's comprised of are to be inserted into the main chain and that's it. Current `MainChain`'s interface inserts blocks by reference, so, IMO, to deserialize them as vector of `shared_ptr`s would induce an unneeded overhead. OTOH, when they are collected in `MainChain::TimeTravel`, they are collected as pointers of `block_chain_`, and not copied (because they only need to be collected and serialized, pointers are good for that ... perhaps).

One obvious change (besides ignoring the fact of this overhead existing at all) would be to allow `MainChain` to insert blocks by shared pointer; then they can be deserialized as ready objects on the heap and inserted as such, without any extra copies or allocations; but this looks like a subject for a different, much less PR, where this templating could be eliminated successfully.

Reply:
Let's just open a jira ticket to do this refactoring in subsequent PRs. It seems that it should be fine to use shared pointers to blocks outside of the main chain (in block coordinator and main_chain_rpc), and use a std::move to add blocks to the main chain.